### ISSUE TITLE
dual-homed issue with openvpn udp connections

### ISSUE THREAD

**Issue by gvde** [eduvpn/vpn-server-node#36]
There seems to be an issue with openvpn servers running with two IP addresses (one public, one internal) and udp connections. If the listen command does not give an IP address it won't connect as the openvpn server seems to set its source address based on the default gateway in the main routing table.

Client sends UDP packet to public IP of server.
Server responds with local IP as source address.

To solve this there is an option multihome which handles these cases.

It would be good if the server config automatically does this as this is probably a common pitfall...

**Comment by fkooman** [comment]
```
       --multihome
              Configure  a  multi-homed  UDP  server.  This option needs to be
              used when a server has more than one IP address  (e.g.  multiple
              interfaces, or secondary IP addresses), and is not using --local
              to force binding to one specific address only.  This option will
              add some extra lookups to the packet path to ensure that the UDP
              reply packets are always sent from the address that  the  client
              is  talking  to.  This is not supported on all platforms, and it
              adds more processing, so it's not enabled by default.

              Note: this option is only relevant for UDP servers.

              Note 2: if you do  an  IPv6+IPv4  dual-stack  bind  on  a  Linux
              machine   with   multiple  IPv4  address,  connections  to  IPv4
              addresses will not work right on kernels  before  3.15,  due  to
              missing  kernel support for the IPv4-mapped case (some distribu‐
              tions have ported this to earlier kernel versions, though).
```

The strange this is that until now we never had any problems with this, even though there were for sure machines with multiple interfaces involved, I think of of our servers have multiple interfaces...

> set its source address based on the default gateway in the main routing table

what does this mean exactly?

**Comment by gvde** [comment]
> > set its source address based on the default gateway in the main routing table
> 
> what does this mean exactly?

Server receives initial packet from client IP. Looks up main routing table for route to client IP, determines interface to send it there and picks the source IP address from the routing table. At least on Linux.

We use routing policies (ip rule) to make sure that traffic is always routed through the correct interface depending the source IP of the outgoing packet.

For our server we have the internal network as "normal" network, i.e. the default gateway is the internal router.

The server has an additional public IP address (which is also a different firewall zone). Without policy routing outgoing traffic would always go through the internal router (using the public IP address as source address if it's handled properly).

The routing policies route outgoing packets with the public IP source address through the public gateway instead of the internal one to make sure packets go in and out through the same interface.

Without the multihome option openvpn does not set the source address and due to that the policies won't apply and it uses the main routing table.

**Comment by fkooman** [comment]
There's no other way to do this? We also use source routing to route the (public IPv4/IPv6) VPN client ranges through another interface, but the default gateway is only changed for that particular range, not for the VPN server itself and that works great without any issues...

**Comment by gvde** [comment]
> There's no other way to do this? We also use source routing to route the (public IPv4/IPv6) VPN client ranges through another interface, but the default gateway is only changed for that particular range, not for the VPN server itself and that works great without any issues...

So the "main" IP address of your VPN server is the public IP (it's the interface with the default route in the main routing table) and the private IP is secondary? That means, for example, SSH connections to the VPN server go through the public IP.

We have the exact opposite. The server is internal (in particular internal traffic like SSH goes through here) and it has an additional public IP for the HTTP and OpenVPN connections...

**Comment by fkooman** [comment]
> So the "main" IP address of your VPN server is the public IP (it's the interface with the default route in the main routing table) and the private IP is secondary?

Yes. But I wouldn't use the word "secondary", it is just another interface, nothing special.

> That means, for example, SSH connections to the VPN server go through the public IP.

No, there is a "management VLAN" for SSH etc.
 
> it has an additional public IP for the HTTP and OpenVPN connections...

I guess that's one approach... we try to keep things as simple as possible...

**Comment by efef** [comment]
> There seems to be an issue with openvpn servers running with two IP addresses (one public, one internal) and udp connections. If the listen command does not give an IP address it won't connect as the openvpn server seems to set its source address based on the default gateway in the main routing table.

Hi @gvde thanks for the notice. Do you have a reference link related to this issue? I assume this topic has been addressed on the OpenVPN mailing-list?

**Comment by fkooman** [comment]
We've had some more internal discussion about this, actually mostly me explaining what the actual problem was ;-)

The main reason for being against adding this is to keep the number of (configuration) options as low as possible. There are already too many as it stands. Adding the option to augment the configuration using a "custom" field also seems to have the potential to lead to disaster when admins will start adding options there as well overriding (other) behavior resulting in something that can no longer reasonably be supported...

I'd make it the default, but it has a performance impact according to the man page, however I never quantified that, it may not be that big of a deal... Do you have any measurements regarding this?

As nobody else has this kind of "strange" routing, I'm not really sure what to do here... 

1. Do nothing, you can add "multihome" to your OpenVPN config every time after running `vpn-server-node-server-config`, possibly using a shell script that runs vpn-server-node-server-config and `tee`s it automatically;
2. Add the (undocumented, unsupported) `enableMultihome` option to vpn-server-api/vpn-server-node to allow you to set this option;
3. Add a "free form" augment for configuration options added to the OpenVPN config

Option 3 is clearly not the way to go ;-) I'm not sure what to do...

**Comment by gvde** [comment]
I guess, ideally, if the server where eduvpn gets installed has multiple IP addresses or multiple interfaces the admin should be able to configure which IP address(es)/interface(s) are used for incoming client connections and which is for the internal side. That way you could automatically add necessary ip rules & routes to route the traffic correctly and also bind the openvpn server only to the required interfaces and even configure iptables, if necessary.

Of course, you would need to know or specify the correct default gateway for both sides.

The way you do it with ip rules based on the source IPs of the VPN clients  requires you to adjust the rules if you add another configuration to the server using a new subnet, doesn't it?

**Comment by fkooman** [comment]
> I guess, ideally, if the server where eduvpn gets installed has multiple IP addresses or multiple interfaces the admin should be able to configure which IP address(es)/interface(s) are used for incoming client connections and which is for the internal side. That way you could automatically add necessary ip rules & routes to route the traffic correctly and also bind the openvpn server only to the required interfaces and even configure iptables, if necessary.

This is already [possible](https://github.com/eduvpn/documentation/blob/v2/PROFILE_CONFIG.md#options) with the `listen` option in `/etc/vpn-server-api/config.php`. You can make OpenVPN just listen on the external interface. The problem is that you'll have to either provide an IPv4 or IPv6 address, it can't be both. Only by not specifying it, it will listen on both IPv4 and IPv6. I don't know if this would help?

**Comment by gvde** [comment]
> This is already [possible](https://github.com/eduvpn/documentation/blob/v2/PROFILE_CONFIG.md#options) with the `listen` option in `/etc/vpn-server-api/config.php`. You can make OpenVPN just listen on the external interface. The problem is that you'll have to either provide an IPv4 or IPv6 address, it can't be both. Only by not specifying it, it will listen on both IPv4 and IPv6. I don't know if this would help?

No the listen only affects on what interface the openvpn server listens to. It does not affect the traffic inside the VPN tunnel.

The problem is if you have a simple dual-homed server setup (i.e. two interfaces each with an IP address in a different network, one external and one internal, and one default gateway) you need to set up some additional rules/routing to get the different routing for the openvpn traffic and the tunneled traffic.

Let's say eth0 is connected to an external/public network and eth1 is connected to the internal network. Users connect via the external network to the external IP and then want internal network access (and possibly all internet traffic going through the internal network).

Scenario 1: The default gateway is configured on eth0. Thus, the traffic in the tunnel needs different routing to use the gateway on eth1.

Scenario 2: The default gateway is configured on eth1. Now, you need a different gateway address for the openvpn server traffic on eth0 to make sure you have symmetric routing and openvpn server traffic enters and leaves through eth0. In that case you also have to set the multihome option because otherwise UDP response packets from the openvpn server won't have a source IP address set when they enter the routing process.

Either way, simply having two IP addresses and one default gateway as it is set up by basic interface configuration won't be enough...

**Comment by fkooman** [comment]
> No the listen only affects on what interface the openvpn server listens to. It does not affect the traffic inside the VPN tunnel.

Yes, exactly. For that the traffic inside the VPN you use source routing, to route it to your internal IP.

> The problem is if you have a simple dual-homed server setup (i.e. two interfaces each with an IP address in a different network, one external and one internal, and one default gateway) you need to set up some additional rules/routing to get the different routing for the openvpn traffic and the tunneled traffic.

True.

> Scenario 1: The default gateway is configured on eth0. Thus, the traffic in the tunnel needs different routing to use the gateway on eth1.

This is what we use a lot, doesn't (seem to) need `multihome`.

> Scenario 2: The default gateway is configured on eth1. Now, you need a different gateway address for the openvpn server traffic on eth0 to make sure you have symmetric routing and openvpn server traffic enters and leaves through eth0. In that case you also have to set the multihome option because otherwise UDP response packets from the openvpn server won't have a source IP address set when they enter the routing process.

How can you get symmetric routing here? Also some kind of source? destination? routing? Are you sure `multihome` actually works in this scenario? It seems like a really complicated thing to do... And what is the benefit of doing things like this? Why would anyone want to do this?

**Comment by gvde** [comment]
> > Scenario 1: The default gateway is configured on eth0. Thus, the traffic in the tunnel needs different routing to use the gateway on eth1.
> 
> This is what we use a lot, doesn't (seem to) need `multihome`.

No. It doesn't need multihome because the default route of the main routing table assigns the correct source IP address to the packets. But you still need to add source routing for the traffic inside the VPN to use the internal gateway. What do you use for the rule? "from 10.x.x.x/xx" with 10.x.x.x being the IP subnet assigned to the VPN clients? That would mean you have to adjust those rules if your add/change new IP subnets.

(Neglecting the asymmetric routing that happens when an external IP address addresses the internal IP address of the server unless you use source routing here, too...)

> > Scenario 2: The default gateway is configured on eth1. Now, you need a different gateway address for the openvpn server traffic on eth0 to make sure you have symmetric routing and openvpn server traffic enters and leaves through eth0. In that case you also have to set the multihome option because otherwise UDP response packets from the openvpn server won't have a source IP address set when they enter the routing process.
> 
> How can you get symmetric routing here? Also some kind of source? destination? routing? Are you sure `multihome` actually works in this scenario? It seems like a really complicated thing to do... And what is the benefit of doing things like this? Why would anyone want to do this?

It's not more complicated than the other scenario. On the contrary: for the routing you just need a rule based on the external source IP and a simple routing table associated with it. For TCP and ICMP that's enough as that sets the source address correctly. Now for UDP the openvpn server doesn't set the source address which means it won't work unless you set the multihome option to force the server to set the source address. But then, everything is symmetric.

In short: eth0 IP x.x.1.100/24 (gw x.x.1.1). eth1 IP x.x.2.100/24 default gw x.x.2.1

ip rule add from x.x.1.100/32 lookup ext
ip route add table ext x.x.1.0/24 dev eth0 src x.x.1.100
ip route add table ext default via x.x.1.1 dev eth0

For the traffic inside the tunnel the standard routing table puts everything into the internal network.

For the openvpn server traffic the rule and the multihome option make sure that the server traffic stays on the external network. There won't be any asymmetric routing and you won't have to adjust your routing depending on what IP addresses you assign to your vpn clients.

But of course, for both scenarios you'll have to add rule/routes in addition to the basic IP interface configuration.
